File: 371.sum-of-two-integers.py
#
# @lc app=leetcode id=371 lang=python3
#
# [371] Sum of Two Integers
#

# @lc code=start


class Solution:
    def getSum(self, a: int, b: int) -> int:

        add = lambda a, b: add(a ^ b, (a & b) << 1) if b else a
        inv = lambda x: add(~x, 1)

        if a * b < 0:
            if a > 0:
                a, b = b, a
            ia = inv(a)
            if ia == b:
                return 0
            if ia < b:
                ib = inv(b)
                return inv(add(ia, ib))
        return add(a, b)


# @lc code=end

File: 371.sum-of-two-integers.rs
/*
 * @lc app=leetcode id=371 lang=rust
 *
 * [371] Sum of Two Integers
 */

// @lc code=start
impl Solution {
    pub fn get_sum(a: i32, b: i32) -> i32 {
        fn add(a: i32, b: i32) -> i32 {
            if b == 0 {
                return a;
            }
            add(a ^ b, (a & b) << 1)
        }
        return add(a, b);
    }
}
// @lc code=end

